.github/workflows/go-tests.yml

```diff
@@ -103,6 +103,7 @@ jobs:
             --junitfile "$TEST_RESULTS_DIR/gotestsum-report.xml" -- \
             -tags="$GOTAGS" -p 2 \
             -cover -coverprofile=coverage.txt \
+            -race \
```

Comment:
Critical aspect - this enables the race detector in CI!

Reply:
>  cost of race detection varies by program, but for a typical program, memory usage may increase by 5-10x and execution time by 2-20x

Reply:
Basically, we need to be ready for CI runs to be quite a bit slower. I'd suggest perhaps not running -race to block the PR flow, maybe on merges to main?

Reply:
In practice, it looks like this brings us from ~9 mins to ~10 mins

CI test summary without -race: https://github.com/hashicorp/waypoint/actions/runs/4991989926
And with -race: https://github.com/hashicorp/waypoint/actions/runs/4994064690?pr=4708

And this isn't the slowest check! The tests currently include a dev windows build that takes 20 minutes.

The downside of running this on main only is that, if someone introduces a race bug, they won't notice until after the merge, which will add some confusion. I think the earlier we catch these race conditions, the less time consuming it will be to fix!

I can imagine us introducing tests in the future though that are much slower in race mode though, and we might not realize that we could save significant time in PR test runs by disabling race. 

With all that in mind, wdyt?

Reply:
Ok, good justification. Sounds good to me to use it everywhere.

Reply:
Worth a try at least! Will plan to revert if it becomes a problem.